File: corax/commands.cpp
#include "commands.h"

Commands::Commands(const W::Address& address, QObject* parent):
    M::Vocabulary(address, parent),
    commands(new Map())
{
}

Commands::~Commands()
{
    Map::iterator beg = commands->begin();
    Map::iterator end = commands->end();
    
    for (; beg != end; ++beg) {
        Command* cmd = beg->second;
        if (cmd->enabled) {
            removeHandler(cmd->handler);
        }
        delete cmd->handler;
        delete cmd;
    }
    
    delete commands;
}

void Commands::addCommand(const W::String& key, W::Handler* handler, const W::Vocabulary& args)
{
    Map::const_iterator itr = commands->find(key);
    if (itr != commands->end()) {
        throw 1;
    }
    Command* cmd = new Command{key, handler, args, false};
    commands->insert(std::make_pair(cmd->name, cmd));
}

void Commands::enableCommand(const W::String& key, bool value)
{
    Map::const_iterator itr = commands->find(key);
    if (itr == commands->end()) {
        throw 2;
    }
    
    Command* cmd = itr->second;
    if (cmd->enabled != value) {
        if (value) {
            enableCommand(cmd);
        } else {
            disableCommand(cmd);
        }
    }
}

void Commands::enableCommand(Commands::Command* cmd)
{
    addHandler(cmd->handler);
    cmd->enabled = true;
    
    W::Vocabulary* vc = new W::Vocabulary;
    vc->insert(u"address", cmd->handler->getAddress());
    vc->insert(u"argumanets", cmd->arguments);
    insert(cmd->name, vc);
}

void Commands::disableCommand(Commands::Command* cmd)
{
    removeHandler(cmd->handler);
    cmd->enabled = false;
    erase(cmd->name);
}

File: magnus/core/commands.js
"use strict";

var ModelVocabulary = require("../lib/wModel/vocabulary");

var Vocabulary = require("../lib/wType/vocabulary");
var String = require("../lib/wType/string");

var Commands = ModelVocabulary.inherit({
    "className": "Commands",
    "constructor": function(address) {
        ModelVocabulary.fn.constructor.call(this, address);
        
        this._commands = global.Object.create(null);
    },
    "destructor": function() {
        for (var key in this._commands) {
            var cmd = this._commands[key];
            if (cmd.enabled) {
                this._removeHandler(cmd.handler);
            }
            cmd.name.destructor();
            cmd.handler.destructor();
            cmd.arguments.destructor();
            delete this._commands[key];
        }
        
        ModelVocabulary.fn.destructor.call(this);
    },
    "addCommand": function(key, handler, args) {
        if (this._commands[key]) {
            throw new Error("Command with this key already exist");
        }
        this._commands[key] = {
            name: new String(key),
            handler: handler,
            arguments: args,
            enabled: false
        }
    },
    "_disableCommand": function(cmd) {
        this._remveHandler(cmd.handler);
        cmd.enabled = false;
        this.erase(cmd.name);
    },
    "enableCommand": function(key, value) {
        var cmd = this._commands[key];
        
        if (!cmd) {
            throw new Error("An attempt to access non existing command");
        }
    
        if (cmd.enabled !== value) {
            if (value) {
                this._enableCommand(cmd);
            } else {
                this._disableCommand(cmd);
            }
        }
    },
    "_enableCommand": function(cmd) {
        this._addHandler(cmd.handler);
        cmd.enabled = true;
        
        var vc = new Vocabulary();
        vc.insert("address", cmd.handler.address.clone());
        vc.insert("argumanets", cmd.arguments.clone());
        
        this.insert(cmd.name.clone(), vc);
    }
});

module.exports = Commands;
